src/coreclr/interpreter/compiler.cpp

```diff
@@ -4612,37 +4911,643 @@ void InterpCompiler::EmitCall(CORINFO_RESOLVED_TOKEN* pConstrainedToken, bool re
+            suspendData->asyncMethodReturnTypePrimitiveSize = 4;
+            break;
+        case CORINFO_TYPE_LONG:
+        case CORINFO_TYPE_ULONG:
+            suspendData->asyncMethodReturnTypePrimitiveSize = 8;
+            break;
+        case CORINFO_TYPE_FLOAT:
+            suspendData->asyncMethodReturnTypePrimitiveSize = 4;
+            break;
+        case CORINFO_TYPE_DOUBLE:
+            suspendData->asyncMethodReturnTypePrimitiveSize = 8;
+            break;
+        case CORINFO_TYPE_BOOL:
+            suspendData->asyncMethodReturnTypePrimitiveSize = 1;
+            break;
+        case CORINFO_TYPE_CHAR:
+            suspendData->asyncMethodReturnTypePrimitiveSize = 2;
+            break;
+        case CORINFO_TYPE_BYREF:
+            BADCODE("ByRef return types not supported for async methods");
+            break;
+        case CORINFO_TYPE_NATIVEINT:
+        case CORINFO_TYPE_NATIVEUINT:
+            suspendData->asyncMethodReturnTypePrimitiveSize = sizeof(void*);
+            break;
+
+        case CORINFO_TYPE_VOID:
+            suspendData->asyncMethodReturnType = NULL;
+            suspendData->asyncMethodReturnTypePrimitiveSize = 0;
+            break;
+        case CORINFO_TYPE_PTR:
+            suspendData->asyncMethodReturnTypePrimitiveSize = sizeof(void*);
+            break;
+        case CORINFO_TYPE_REFANY:
+            BADCODE("TypedReference return types not supported for async methods");
+            break;
+
+        case CORINFO_TYPE_VAR:
+            BADCODE("VAR return types should have been translated before we get here on async methods");
+            break;
+        default:
+            // Other types are not supported
+            BADCODE("Unsupported async method return type");
+            break;
+    }
+
+    // Step 7: Emit the INTOP_HANDLE_CONTINUATION instruction
+
+    CorInfoHelpFunc helperFuncForAllocatingContinuation = CORINFO_HELP_ALLOC_CONTINUATION;
+    int handleContinuationOpcode = INTOP_HANDLE_CONTINUATION;
+    int32_t varGenericContext = -1;
+
+    if (kindForAllocationContinuation.needsRuntimeLookup)
+    {
+        handleContinuationOpcode = INTOP_HANDLE_CONTINUATION_GENERIC;
+        switch (kindForAllocationContinuation.runtimeLookupKind)
+        {
+        case CORINFO_LOOKUP_CLASSPARAM:
+            helperFuncForAllocatingContinuation = CORINFO_HELP_ALLOC_CONTINUATION_CLASS;
+            varGenericContext = getParamArgIndex();
+            break;
+        case CORINFO_LOOKUP_THISOBJ:
+        {
+            helperFuncForAllocatingContinuation = CORINFO_HELP_ALLOC_CONTINUATION_CLASS;
+            AddIns(INTOP_LDIND_I);
+            m_pLastNewIns->data[0] = 0; // The offset is 0 for the this pointer
+            m_pLastNewIns->SetSVar(getParamArgIndex());
+            PushInterpType(InterpTypeI, NULL);
+            varGenericContext = m_pStackPointer[-1].var;
+            m_pLastNewIns->SetDVar(varGenericContext);
+            m_pStackPointer--;
+            break;
+        }
```

Comment:
Is this needed?
On the JIT side we rely on the on-entry "this" object to be considered live and thus captured normally as part of the continuation.
Hence in that case I don't think a separate explicit keep alive slot is necessary.